packages/postcss-tape/src/index.ts

```diff
@@ -0,0 +1,344 @@
+					}
+				} catch (_) {
+					hasErrors = true;
+
+					if (process.env.GITHUB_ACTIONS) {
+						console.log(formatGitHubActionAnnotation(
+							`${testCaseLabel}\n\nresult was not parsable with PostCSS.`,
+							'error',
+							{ file: expectFilePath, line: 1, col: 1 },
+						));
+					} else {
+						console.error(`\n${testCaseLabel}\n\nresult was not parsable with PostCSS.\n\n${dashesSeparator}`);
+					}
+				}
+			}
+
+			// Assert result with oldest supported PostCSS.
+			// Check against the actual result to avoid duplicate warnings.
+			//
+			// NOTE:
+			// The oldest version of PostCSS we support at this time.
+			// This does not need to be a different version than the latest version of PostCSS.
+			// There is no system behind setting this.
+			// It is here to allow testing a specific older version, if parts of the community can't update yet.
+			{
+				const resultFromOldestPostCSS = await postcssOldestSupported(plugins as Array<AcceptedPlugin>).process(input, {
+					from: testFilePath,
+					to: resultFilePath,
+					map: {
+						inline: false,
+						annotation: false,
+					},
+				});
+
+				try {
+					assert.strictEqual(resultFromOldestPostCSS.css.toString(), resultString);
+				} catch (err) {
+					hasErrors = true;
+
+					if (process.env.GITHUB_ACTIONS) {
+						console.log(formatGitHubActionAnnotation(
+							'testing older PostCSS:\n' + formatCSSAssertError(testCaseLabel, testCaseOptions, err, true),
+							'error',
+							{ file: normalizeFilePathForGithubAnnotation(expectFilePath), line: 1, col: 1 },
+						));
+					} else {
+						console.error('testing older PostCSS:\n' + formatCSSAssertError(testCaseLabel, testCaseOptions, err));
+					}
+				}
+			}
+
+			// Assert that warnings have the expected amount.
+			{
+				try {
+					if (result.warnings().length || testCaseOptions.warnings) {
+						assert.strictEqual(result.warnings().length, testCaseOptions.warnings);
+					}
+				} catch (err) {
+					hasErrors = true;
+
+					if (process.env.GITHUB_ACTIONS) {
+						console.log(formatGitHubActionAnnotation(
+							formatWarningsAssertError(testCaseLabel, testCaseOptions, result.warnings().length, testCaseOptions.warnings, true),
+							'error',
+							{ file: normalizeFilePathForGithubAnnotation(expectFilePath), line: 1, col: 1 },
+						));
+					} else {
+						console.error(formatWarningsAssertError(testCaseLabel, testCaseOptions, result.warnings().length, testCaseOptions.warnings));
+					}
+				}
+			}
+		}
+
+		if (hasErrors) {
+			process.exit(1);
+		}
+	};
+}
+
+function normalizeFilePathForGithubAnnotation(filePath: string) {
+	// Any plugin or packages is located in `<workspace>/<package>`;
+	const parts = process.cwd().split('/').slice(-2);
+
+	const workspaces = [
+		'packages',
+		'plugins',
+		'plugin-packs',
+		'cli',
+		'experimental',
+	];
+
+	if (!workspaces.includes(parts[0])) {
+		throw new Error('PostCSS Tape was intended to be run from <workspace>/package');
+	}
```

Comment:
Perhaps this is down the line 😂  but what about testing with the rest of the environments?

Reply:
My comment is perhaps confusing and should be updated to `<workspace>/<package>`

I might have been a bit lazy here, but there is currently no good way (afaik) to get a file path relative to an npm workspace root.

Since this can differ widely between local setups and also CI, I look for a pattern with a prefix of the known npm workspaces.

For github annotations we have to log file paths relative to the repo root which is the same as the workspace root.

From `/Users/<you>/some-projects-dir/postcss-plugins/plugins/postcss-is-pseudo-class/test/basic.expect.css` I want to get just `plugins/postcss-is-pseudo-class/test/basic.expect.css`.

This heuristic limits us in a way.
We can't do this without updating the code here :

- `plugins/plugins-for-color/css-color-4`
- `plugins/plugins-for-color/css-color-5`

Currently we don't have a need for this hierarchy, so I am hoping it's fine for now :)

Reply:
_my comments above can be ignored_

----

Discussed : current setup didn't work in node 12/14
I had forgotten about this.

Are we ok with doing this for now?

```js
import postcssTape from '../../packages/postcss-tape/dist/index.mjs';
```

It works in all node versions and doesn't require publishing the test suite.

I would prefer this as I want to have the freedom to break the API when needed without needing to consider other users.

Reply:
I'm completely ok with this approach and also helps with my point above (sharing code)

One other option is https://www.npmjs.com/package/module-alias

Reply:
I added a note in the README about how to clean this up after node 20 is released.
That should make it easy to move fully to workspaces then.